File: apps/ios/Packages/Model/Sources/Model/ServerAddress.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

public struct ServerAddress: Sendable, Equatable, Hashable, CustomStringConvertible {
    public let url: URL
    public var description: String { "[SERVER]" }
    public var isPlainHTTP: Bool { url.scheme == "http" }
    public var apiBase: URL { url.appending(path: "api/v1") }

    public init?(_ input: String) {
        let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, !text.contains(where: { $0.isWhitespace || $0 == "\\" }) else { return nil }
        let value = text.contains("://") ? text : "https://" + text
        guard var parts = URLComponents(string: value),
            let scheme = parts.scheme?.lowercased(), ["https", "http"].contains(scheme),
            let host = parts.host, !host.isEmpty, parts.user == nil, parts.password == nil,
            parts.query == nil, parts.fragment == nil,
            parts.port.map({ (1...65535).contains($0) }) ?? true,
            !parts.percentEncodedPath.lowercased().contains("%2f"),
            !parts.percentEncodedPath.lowercased().contains("%5c"),
            !parts.path.split(separator: "/").contains(where: { $0 == "." || $0 == ".." })
        else { return nil }
        parts.scheme = scheme
        parts.host = host.lowercased()
        while parts.path.hasSuffix("/") { parts.path.removeLast() }
        guard let url = parts.url else { return nil }
        self.url = url
    }
}

File: apps/android/core/model/src/main/kotlin/app/gauja/core/model/ServerAddress.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.model

import java.net.URI

@JvmInline
value class ServerAddress private constructor(val value: String) {
    val apiBase: String
        get() = value.trimEnd('/') + "/api/v1/"

    val isPlainHttp: Boolean
        get() = value.startsWith("http://")

    override fun toString(): String = "[SERVER]"

    companion object {
        fun parse(input: String): ServerAddress? {
            val trimmed = input.trim()
            if (trimmed.isEmpty() || trimmed.any { it.isWhitespace() || it == '\\' }) return null
            val value = if ("://" in trimmed) trimmed else "https://$trimmed"
            val uri =
                try {
                    URI(value)
                } catch (_: java.net.URISyntaxException) {
                    return null
                }
            if (
                uri.scheme?.lowercase() !in setOf("https", "http") ||
                    uri.host.isNullOrBlank() ||
                    uri.rawUserInfo != null ||
                    uri.rawQuery != null ||
                    uri.rawFragment != null ||
                    uri.port !in -1..65535 ||
                    uri.port == 0 ||
                    uri.normalize().rawPath != uri.rawPath ||
                    uri.path.split('/').any { it == "." || it == ".." } ||
                    Regex("(?i)%2f|%5c").containsMatchIn(uri.rawPath)
            )
                return null
            return ServerAddress(
                URI(
                        uri.scheme.lowercase(),
                        null,
                        uri.host.lowercase(),
                        uri.port,
                        uri.path.trimEnd('/'),
                        null,
                        null,
                    )
                    .toASCIIString()
            )
        }
    }
}
